src/profile/ProfileForm.tsx

```diff
@@ -37,585 +25,405 @@ interface ChangedUser {
+  useEffect(() => {
+    register({
+      name: "is_editor",
+    });
+  }, [register]);
+
+  // we need a new password if we're creating a new user
+  const newPasswordRequired = props.user === undefined;
+  // require confirm if we're creating a new editor or changing password/role
+  const passwordConfirmRequired =
+    (newPasswordRequired && props.context.getValues().is_editor) ||
+    (props.user &&
+      (props.context.getValues().password ||
+        props.context.getValues().is_editor !== props.user?.is_editor));
+
+  const [validateUserNameDebounced] = useDebouncedCallback(
+    validateUsernameAvailable,
+    250
+  );
 
-        if (value.length < 8) {
-          newErrors.push("USER.PASSWORD.TOO_SHORT.8");
-        }
+  const [showPassword, setShowPassword] = useState<boolean>(false);
 
-        if (/^\d*$/.test(value)) {
-          newErrors.push("USER.PASSWORD.ENTIRELY_NUMERIC");
-        }
+  const submit = async (vals: ProfileFormVals) => {
+    const cleaned = cleanFormData(vals);
 
-        dispatch({
-          type: "set data",
-          data: { password: value },
-          errors: { status_code: 0, password: newErrors },
-        });
-        break;
+    const editingMe = props.user?.username === auth.user?.username;
+    try {
+      if (props.user === undefined) {
+        await auth.post<ProfileFormVals>("users/", cleaned);
+        setSuccessMessage(`User ${vals.username} created successfully.`);
+        // refresh the form so we can add a new user
+        props.context.reset();
+      } else {
+        await auth.patch<ProfileFormVals>(
+          editingMe ? "me/" : `users/${vals?.username}/`,
+          cleaned,
+          editingMe
+        );
+        setSuccessMessage(`User ${vals.username} saved successfully.`);
       }
-    }
-  };
-
-  const onSubmit = (evt?: React.FormEvent<HTMLFormElement>) => {
-    evt?.preventDefault();
-
-    if (
-      (Object.keys(state.changedUser) as Array<keyof ChangedUser>).every(
-        (k) =>
-          state.changedUser[k] === undefined ||
-          (k === "is_editor" && state.changedUser[k] === user?.is_editor)
-      )
-    ) {
-      return Promise.resolve();
-    }
-
-    if (Object.values(state.errors).flat().length > 1) {
-      dispatch({
-        type: "set general error",
-        errors: [ERRORS.FORMS.NOT_YET_VALID],
-      });
-      return Promise.reject();
-    }
-
-    dispatch({ type: "is loading" });
-
-    return (user === undefined
-      ? auth.post<ChangedUser>("users/", state.changedUser)
-      : auth.patch<ChangedUser>(
-          user === auth.user ? "me/" : `users/${user?.username}/`,
-          state.changedUser,
-          user === auth.user
-        )
-    ).then(
-      () => {
-        dispatch({ type: "done loading success" });
-      },
-      (err: UserAPIError | string) => {
-        dispatch({ type: "done loading error", errors: err });
+      setGeneralErrors(undefined);
```

Comment:
We should also reset showPassword to be false here, so users don't get a nasty surprise.

Reply:
https://github.com/UWmathNEWS/slugline-web/pull/13/commits/24108681801fdd0e22ee5fd2198b3ffaac4bca46